blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-
-        if whence == io.SEEK_SET:
-            self.position = offset
-        elif whence == io.SEEK_CUR:
-            self.position += offset
-        elif whence == io.SEEK_END:
-            self.position = self.object.content_length + offset
-        else:
-            raise ValueError(
-                f'whence must be one of {self.WHENCE_LIST}, not {whence}.'
-            )
-        if self.position < 0:
-            self.position = old_position
-            raise ValueError(
-                f'seek would result in negative position {self.position}'
-            )
-
-        return self.position
-
-    def seekable(self) -> bool:
-        return True
-
-    def read(self, size: int = -1) -> bytes:
-        # GetObject will be unhappy if we try to read from beyond the
-        # end of the object.
-        if self.position >= len(self):
-            return b''
-
-        if size == -1:
-            range = f'bytes={self.position}-'
-            self.seek(0, io.SEEK_END)
-        else:
-            # But it's fine to specify a range that goes beyond the end
-            # if it starts before.
-            # Subtract one because the range is inclusive.
-            range = f'bytes={self.position}-{self.position + size - 1}'
-            self.seek(size, io.SEEK_CUR)
-
-        # For some reason boto3 type stubs get the type wrong here.
-        # Explanation:
-        #   a GetObject response has a Body parameter which in normal
-        # circumstances is always of type StreamingBody.
-        # Code in botocore.endpoint is responsible for creating the
-        # StreamingBody object. But the result could instead be a string
-        # (if there is an error) or bytes (if handling a different
-        # parameter.)
-        response = self.object.get(Range=range)
-        status = response['ResponseMetadata']['HTTPStatusCode']  # type: ignore
-        if status > 300:
-            raise BlobError(f'Received status code {status} '
-                            f'when getting blob {self.object}')
-        # Make the assumption that the typecheck ignore is making
-        # explicit
-        assert isinstance(response['Body'], StreamingBody), (
-            f"Response from S3 of type {type(response['Body'])}, "
-            f"not StreamingBody.")
-
-        return response['Body'].read()  # type: ignore
-
-    def readinto(self, b: bytearray) -> int:
-        to_read = min(len(self) - self.position, len(b))
-        b[:to_read] = self.read(to_read)
-        return to_read
-
-    def readable(self) -> bool:
-        return True
-
-    def tell(self) -> int:
-        return self.position
-
-    def __repr__(self) -> str:
-        return f'<{type(self).__name__} wrapping {self.object}>'
-
-    def __len__(self) -> int:
-        return self.object.content_length
-
-
-WriterCallback = Callable[[str], None]
-@inherit_docstrings
-class BlobWriter(io.BufferedIOBase):
-    """A file-like object wrapping an S3 blob with an automatic key.
```

Comment:
IMO it's worth including that 'automatic key' = hash of file contents.

Reply:
I think I actually originally included this but then removed it since I think I remember Michael saying that the specific way of picking the key automatically was intended to be an implementation deal. Actually I had also wondered to myself if a GUID 4 would be better since you wouldn't have to read the whole thing to calculate a hash :P